### PR TITLE
feat(project): introduce path for various operations

### PR DESCRIPTION
**Pull request by mjenny**
- `moveFile()`: allow to move files using paths instead of IDs
- `moveFolder()`: allow to move folders using paths instead of IDs
 - `createFolder()`: allow to create folders using paths instead of IDs
- `deleteFolder()`: allow to delete folders using paths instead of IDs
- `createFile()`: allow to create files using paths instead of IDs
- `deleteFile()`: allow to delete files using paths instead of IDs
- `createFileLink()`: allow to create file links using paths instead of IDs

### CODE REVIEW

File: java-core/src/main/java/com/pcloud/sdk/internal/RealApiClient.java [hunk 1]
```diff
@@ -172,10 +172,12 @@ public RemoteFolder adapt(Response response) throws IOException, ApiError {
...
     @Override
     public Call<RemoteFolder> listFolder(String path, boolean recursively) {
-
+        if (path == null) {
+            throw new IllegalArgumentException("Path argument cannot be null.");
+        }
         HttpUrl.Builder urlBuilder = apiHost.newBuilder()
                 .addPathSegment("listfolder")
-                .addEncodedQueryParameter("path", String.valueOf(path))
+                .addEncodedQueryParameter("path", path.isEmpty() ? "/" : path)
```
Comment (mjenny) [inline java-core/src/main/java/com/pcloud/sdk/internal/RealApiClient.java:180]: Since all folders are loaded without trailing slash empty string should result in loading the root folder.

If this is something you want to have fixed on the server side, let me know.

File: java-core/src/main/java/com/pcloud/sdk/internal/RealApiClient.java [hunk 2]
```diff
@@ -350,7 +409,27 @@ public Boolean adapt(Response response) throws IOException, ApiError {
...
+            public FileLink adapt(Response response) throws IOException, ApiError {
+                return getAsFileLink(response);
+            }
+        });
+
+    }
+
+    @Override
+    public Call<FileLink> createFileLink(String path, DownloadOptions options) {
+        if (path == null) {
```
Comment (georgi-neykov-hub) [inline java-core/src/main/java/com/pcloud/sdk/internal/RealApiClient.java:425]: Non-null but empty `path` arguments should also be treated as an illegal argument, so maybe it's a good idea to add a helper function that does the checks and throws with an appropriate message, then replace sanity checks throughout public methods with a `path` argument with the new function.

File: java-core/src/test/java/com/pcloud/sdk/ApiServiceTest.java [hunk 3]
```diff
@@ -87,6 +87,12 @@ public void createFolder_ThrowsOnNullRemoteFolderArgument() throws Exception {
         instance.createFolder(null, null);
     }
 
+    @Test
```
Comment (georgi-neykov-hub) [inline java-core/src/test/java/com/pcloud/sdk/ApiServiceTest.java:90]: It would be good f the tests for the the new path-based methods also test the methods' output, not only the sanity checks presence.

**Review by georgi-neykov-hub — CHANGES_REQUESTED** [review]

File: java-core/src/main/java/com/pcloud/sdk/internal/RealApiClient.java (hunk 1 above)
Comment (mjenny) [reply java-core/src/main/java/com/pcloud/sdk/internal/RealApiClient.java:180]: Let's say you want to `listFolder` `/Test/Sub Folder`. There is no trailing slash meaning `listFolder("/Test/Sub Folder/")` results in an error.

To keep this consistent, empty string should reference the root folder since `listFolder("/")` is the trailing slash.

File: java-core/src/main/java/com/pcloud/sdk/internal/RealApiClient.java (hunk 2 above)
Comment (mjenny) [reply java-core/src/main/java/com/pcloud/sdk/internal/RealApiClient.java:425]: Let's say you want to `listFolder` `/Test/Sub Folder`. There is no trailing slash meaning `listFolder("/Test/Sub Folder/")` results in an error.

To keep this consistent, empty string should reference the root folder since `listFolder("/")` is the trailing slash.

File: java-core/src/test/java/com/pcloud/sdk/ApiServiceTest.java (hunk 3 above)
Comment (mjenny) [reply java-core/src/test/java/com/pcloud/sdk/ApiServiceTest.java:90]: Agree. Will implement them.

File: java-core/src/main/java/com/pcloud/sdk/internal/RealApiClient.java (hunk 2 above)
Comment (georgi-neykov-hub) [reply java-core/src/main/java/com/pcloud/sdk/internal/RealApiClient.java:425]: If you need to access the content of `/Test/Sub Folder`, why not call the method with the same path without adding the trailing`/`?  Doing a `listfolder` on the parent would also return a `/Test/Sub Folder` for the same child folder.

Comment (mjenny) [reply java-core/src/main/java/com/pcloud/sdk/internal/RealApiClient.java:425]: Yeah sure, but I'd say `/` to list the root folder is like `/Test/` to list the Test folder which isn't working.

But yeah,  it's an edge case. If this is needed to get the PR approved I'll be happy to change it. No hard feelings. 😄

**Comment by georgi-neykov-hub** [comment]
@mjenny can you unify the null/empty path checks in a single method and apply this to all implementations of methods that accept a `path` parameter?

**Comment by mjenny** [comment]
> 
> 
> @mjenny can you unify the null/empty path checks in a single method and apply this to all implementations of methods that accept a `path` parameter?

Done.

**Comment by georgi-neykov-hub** [comment]
> > @mjenny can you unify the null/empty path checks in a single method and apply this to all implementations of methods that accept a `path` parameter?
> 
> Done.

@mjenny thank you, I will try to take a final look in the next few hours and merge the changes. If everything is OK I will publish an update with the changes by the end of the day.

**Review by georgi-neykov-hub — APPROVED** [review]

**Comment by mjenny** [comment]
Thank you for merging this PR. 👍

**Comment by georgi-neykov-hub** [comment]
@mjenny, there is an update to the SDK that includes the changes.

**Comment by mjenny** [comment]
@georgi-neykov-hub 
Are you going to link 1.4.0 to JCenter? Having the artifact only on Bintray is not optimal.
